app/src/main/java/com/wut/indoornavigation/data/mesh/Mesh.java

```diff
@@ -0,0 +1,398 @@
+            if (f.getNumber() == k) {
+                kFloor = f;
+                break;
+            }
+        }
+
+        return kFloor;
+    }
+
+    private void linkStairsOnFloor(Building building, Graph graph, Floor floor, int floorNumber) {
+        if (stairsVerticesDict.size() == 0) {
+            return;
+        }
+
+        for (int i = 0; i < stairsVerticesDict.get(floorNumber).size(); i++) {
+            Stairs stairs = floor.getStairs().get(i);
+            linkStairOnFloor(building, graph, floorNumber, i, stairs);
+        }
+    }
+
+    private void linkStairOnFloor(Building building, Graph graph, int floorNumber, int i, Stairs stairs) {
+        if (stairs.getEndfloor() != floorNumber) {
+            int k = stairs.getEndfloor();
+
+            if (building.getFloors().size() < k + 1 || stairsVerticesDict.size() < k + 1) {
+                return;
+            }
+
+            Floor kFloor = getFloorByNumber(building, k);
+
+            if (kFloor == null) {
+                return;
+            }
+
+            List<Stairs> endFloorStairs = kFloor.getStairs();
+
+            if (stairs.getEndfloor() == floorNumber) {
+                return;
+            }
+
+            List<Vertex> endFloorStairsGraphVertices = stairsVerticesDict.get(k);
+            int endVertexIndex = findMatchingStairsOnFloor(stairs, endFloorStairs);
+
+            Vertex startVertex = stairsVerticesDict.get(floorNumber).get(i);
+            Vertex endVertex = endFloorStairsGraphVertices.get(endVertexIndex);
+
+            listStairOnFloor(graph, startVertex, endVertex);
+        }
+    }
+
+    private void listStairOnFloor(Graph graph, Vertex startVertex, Vertex endVertex) {
+        if (startVertex.getId() != endVertex.getId()) {
+            if (!graph.containsEdge(startVertex.getId(), endVertex.getId())) {
+                graph.addEdge(new Edge(startVertex, endVertex, EDGE_STAIRS_WEIGHT));
+            }
+        }
+    }
+
+    private int findMatchingStairsOnFloor(Stairs stairs, List<Stairs> endFloorStairs) {
+        int endVertexIndex = -1;
+        for (int j = 0; j < endFloorStairs.size(); j++) {
+            if (endFloorStairs.get(j).getId() == stairs.getId()) {
+                endVertexIndex = j;
+                break;
+            }
+        }
+
+        if (endVertexIndex == -1) {
+            throw new IllegalStateException("This algorithm is bugged as f*ck.");
+        }
+        return endVertexIndex;
+    }
+
+    @Nullable
+    private Vertex processCell(int x, int y, FloorObject[][] enumMap, int floorNumber, boolean[][] visited, Graph graph) {
+        if (visited[x][y]) {
+            return graph.getVertexByCoordinates((float) x / 2, (float) y / 2, floorNumber);
+        }
+
+        visited[x][y] = true;
+        if (enumMap[x][y] == FloorObject.SPACE || enumMap[x][y] == FloorObject.DOOR || enumMap[x][y] == FloorObject.ROOM || enumMap[x][y] == FloorObject.STAIRS || enumMap[x][y] == FloorObject.ELEVATOR) {
```

Comment:
wow, nie latwiej bylo sprawdzic czym nie powinno byc? + jesli juz zdecydowales sie na takiego ifa to wydziel to do oddzielnej metody ktora zwraca boolean i jej nazwa mowi co wlasciwie sprawdzasz